backend/poll.go

```diff
@@ -0,0 +1,180 @@
+type PollAnswer struct {
+	Answer int
+	Votes  int
+}
+
+//holds the questions
+type PollQuestions struct {
+	Questions []string
+}
+
+//holds stored Poll
+type Poll struct {
+	ClientIds   []string
+	PollAnswers []PollAnswer
+}
+
+var questions []string
+var pollQuestions PollQuestions
+
+func InitPollSample() {
+	questions = []string{"Penguins", "Ostriches", "Kiwis", "Wekas"}
+	pollQuestions = PollQuestions{questions}
+	http.HandleFunc(POLL_SAMPLE_PATH+"submit", submitPoll)
+	RegisterSample(CATEGORY_SAMPLE_TEMPLATES+"/poll", handlePoll)
+}
+
+func handlePoll(w http.ResponseWriter, r *http.Request, page Page) {
+	page.Render(w, pollQuestions)
+}
+
+func createPollResult(answers []PollAnswer, message string) PollResult {
+	results := make([]PollEntryResult, 0)
+	for _, value := range answers {
+		results = append(results, PollEntryResult{make([]int, value.Votes), questions[value.Answer]})
+	}
+	return PollResult{results, message}
+}
+
+func submitPoll(w http.ResponseWriter, r *http.Request) {
+	EnableCors(w, r)
+	SetContentTypeJson(w)
+	context := appengine.NewContext(r)
+	pollForm, error := parsePollForm(w, r, context)
+	if error != nil {
+		handleError(error, w)
+		return
+	}
+	pollResult, error := calculatePollResults(w, context, pollForm)
+	if error != nil {
+		handleError(error, w)
+		return
+	}
+	json.NewEncoder(w).Encode(pollResult)
+}
+
+func parsePollForm(w http.ResponseWriter, r *http.Request, context context.Context) (PollForm, error) {
+	answer, answerErr := parseNotEmptyFormValue(r, "answer")
+	answerId, _ := strconv.Atoi(answer)
+	clientId, clientIdErr := parseNotEmptyFormValue(r, "clientId")
+
+	pollForm := PollForm{clientId, answerId}
+
+	error := parseFormErrors([]error{answerErr, clientIdErr})
+	return pollForm, error
+}
+
+func calculatePollResults(w http.ResponseWriter, ctx context.Context, pollForm PollForm) (PollResult, error) {
+	var message string
+	clientId := pollForm.ClientId
+
+	//get the poll
+	pollKey := datastore.NewKey(ctx, "Poll", "poll", 0, nil)
+	var poll Poll
+	//if it's the first time executing, there will be no poll yet
+	datastore.Get(ctx, pollKey, &poll)
+
+	//check if the user has already voted
+	existingClientIds := poll.ClientIds
+	for _, existingClientId := range existingClientIds {
+		if existingClientId == clientId {
+			//return a message to let the user know that has already voted
+			message = "You have already answered this poll. If you want to run this sample again, use an incognito window."
+			//return the existing poll answers and a message
+			return createPollResult(poll.PollAnswers, message), nil
+		}
+	}
+
+	//if the user hasn't already voted, return a welcome message
+	message = "Thanks for answering the poll!"
+	//add the clientId
+	poll.ClientIds = append(existingClientIds, clientId)
+
+	pollExistingAnswers := poll.PollAnswers
+	//get the lastest vote count for the answer
+	answerFound := false
+	for _, pollExistingAnswer := range pollExistingAnswers {
```

Comment:
You could simplify this a bit by storing the vote count in an array -> the question id is the array index of the vote count. The advantage would be that you could replace the for loop with a single array access.